### PR TITLE
receiver_creator: lookup configured observer extensions

### PR DESCRIPTION
**Pull request by jrcamp**
This utilizes `host.GetExtensions` to find observers that have been configured
using `watch_observers`. receiver_creator can be configured to watch zero or
more observers. To support this the receiver templates have been moved under
"receivers" key to be able to differentiate subreceiver config keys from config
keys for the receiver_creator itself.

### CODE REVIEW

File: receiver/receivercreator/factory.go [hunk 1]
```diff
@@ -52,17 +52,22 @@ func (f *Factory) CustomUnmarshaler() component.CustomUnmarshaler {
...
 		c := intoCfg.(*Config)
 
-		// TODO: Change the sub-receivers to be under "receivers" to allow other config for the main receiver-creator receiver.
-		for subreceiverKey := range sourceViperSection.AllSettings() {
-			cfgSection := sourceViperSection.GetStringMap(subreceiverKey + "::config")
+		if err := sourceViperSection.Unmarshal(&c); err != nil {
+			return err
+		}
+
+		receiversCfg := viperSub(sourceViperSection, "receivers")
```
Comment (tigrannajaryan) [inline receiver/receivercreator/factory.go:59]: Can we define "receivers" as a const somewhere close to `Config` and explain what it is?

File: receiver/receivercreator/factory.go [hunk 2]
```diff
@@ -52,17 +52,22 @@ func (f *Factory) CustomUnmarshaler() component.CustomUnmarshaler {
...
+
+		for subreceiverKey := range receiversCfg.AllSettings() {
+			cfgSection := receiversCfg.GetStringMap(subreceiverKey + "::config")
```
Comment (tigrannajaryan) [inline receiver/receivercreator/factory.go:62]: What does "::" do here? Can you add a comment that describes why we are concatenating the key this way?

File: receiver/receivercreator/receiver.go [hunk 3]
```diff
@@ -86,7 +86,39 @@ func (rc *receiverCreator) Start(ctx context.Context, host component.Host) error
...
+				return fmt.Errorf("extension %q in watch_observers is not an observer", watchObserver)
+			}
+			observers[watchObserver] = obs
+		}
+	}
+
+	// Make sure all observers are present before starting any.
+	for _, watchObserver := range rc.cfg.WatchObservers {
+		if observers[watchObserver] == nil {
+			return fmt.Errorf("failed to find observer %q", watchObserver)
```
Comment (tigrannajaryan) [inline receiver/receivercreator/receiver.go:109]:
```suggestion
			return fmt.Errorf("failed to find observer %q in the extensions list", watchObserver)
```

File: receiver/receivercreator/receiver.go [hunk 4]
```diff
@@ -86,7 +86,39 @@ func (rc *receiverCreator) Start(ctx context.Context, host component.Host) error
...
+		}
+	}
+
+	if len(observers) == 0 {
+		rc.logger.Warn("no observers were configured and no subreceivers will be started",
```
Comment (tigrannajaryan) [inline receiver/receivercreator/receiver.go:114]:
```suggestion
		rc.logger.Warn("no observers were configured and no subreceivers will be started. receiver_creator will be disabled",
```

File: receiver/receivercreator/factory.go (hunk 2 above)
Comment (jrcamp) [reply receiver/receivercreator/factory.go:62]: I'll use viperSub as well so it's more clear.

File: receiver/receivercreator/factory.go [hunk 5]
```diff
@@ -52,17 +52,22 @@ func (f *Factory) CustomUnmarshaler() component.CustomUnmarshaler {
...
-		for subreceiverKey := range sourceViperSection.AllSettings() {
-			cfgSection := sourceViperSection.GetStringMap(subreceiverKey + "::config")
+		if err := sourceViperSection.Unmarshal(&c); err != nil {
+			return err
+		}
+
+		receiversCfg := viperSub(sourceViperSection, receiversConfigKey)
+
+		for subreceiverKey := range receiversCfg.AllSettings() {
+			cfgSection := viperSub(receiversCfg, subreceiverKey).GetStringMap("config")
```
Comment (tigrannajaryan) [inline receiver/receivercreator/factory.go:62]: For consistency please also define "config" close to the `Config` type somewhere near `receiversConfigKey` const.

**Review by tigrannajaryan — APPROVED** [review]
LGTM, just one minor comment left.

File: receiver/receivercreator/factory.go (hunk 5 above)
Comment (jrcamp) [reply receiver/receivercreator/factory.go]: Done

**Comment by tigrannajaryan** [comment]
@jrcamp this seems stuck and is not building. Can you touch a file to see if the build starts?

**Pull request merged**